Implement support for Hidden nodes in taproot trees and Fix taproot serde bugs

sanket1729 opened the pull request:
This PR changes/removes the serde implementation for the following types

- TaprootSpendInfo: Removed. This data structure contains derived information for taproot spending that cannot be validated easily. To elaborate, `TaprootSpendInfo` is constructed from a tree, but loses information about the tree structure and maintains handy information like `script_control_block_map`, cached tweaked key, Merkle root etc. 
- TaprootBuillder: Removed. Hard to implement and not very useful.
- TapTree: Modified to check invariants. 
- NodeInfo: Now implements serde with support for Hidden nodes
- ScriptLeaf: Removed serde. Users should not directly construct this. This is just an output iterator item of `TapTree::script_leaves()`

Data structure changes: 

- Introduced `LeafNode`: Supports Hidden leaves. Has serde implemented
- Cleanly separate `TapTree` and `NodeInfo`:  `TapTree` is a full BIP370 compatible tree with no hidden nodes. `NodeInfo` is a tree that can potentially contain hidden leaves.
- Added `NodeInfo::leaf_nodes`: Iterator that iterates over known and hidden leaves of `NodeInfo`.
- Updated `TapTree::script_leaves`: Iterator that is guaranteed to output known leaves.

sanket1729 commented:
Based on #1523

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -1337,7 +1645,7 @@ mod test {
         let builder = builder.add_leaf(3, e.clone()).unwrap();
 
-        let tree_info = builder.finalize(&secp, internal_key).unwrap();
+        let tree_info = builder.clone().finalize(&secp, internal_key).unwrap();
```
No need to clone here.

sanket1729 replied:
The test below uses builder so I needed to conditionally clone under #cfg(serde). But I have rearranged the code so that it's much cleaner now.

tcharding reviewed (dismissed):
ACK 7cf04aa0070e45c6dd322a695239dd43dc0f7d68

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -322,8 +322,6 @@ impl From<&TaprootSpendInfo> for TapTweakHash {
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
 pub struct TaprootBuilder {
```
Yep, IMO no builder should be (de)serializable, builders are purely Rust construct.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -822,6 +820,54 @@ impl TaprootMerkleBranch {
+            type Value = TaprootMerkleBranch;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a valid consensus-encoded taproot leaf version")
```
To be clear, only this one is blocking.

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -822,6 +820,54 @@ impl TaprootMerkleBranch {
+                }
+                TaprootMerkleBranch::from_collection(hashes)
+                    .map_err(|_| serde::de::Error::custom("Merkle branch len must be <= 128"))
+            }
```
Implementing it automatically with `#[serde(into = "Vec<sha256::Hash>")]`

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -596,6 +602,12 @@ impl TapTree {
+    /// Gets the inner [`NodeInfo`] of this tree root.
+    pub fn into_node_info(mut self) -> NodeInfo {
+        // PANICs: Refer to note in `node_info` method above
+        self.0.branch.pop().expect("builder invariant").expect("builder invariant")
```
Indeed, I already am working on this locally. Raised this PR initially to move things forward.

Kixunil commented:
Crap, you pushed in the middle of me reviewing. Do you intend to make more changes?

sanket1729 commented:
I plan to squash the last 4 commits into one after I get access to my dev machine. This should be ready for review otherwise. There are still rough edges, which I can fix after we have #1479 in. I will spend time on it first

sanket1729 commented:
I squashed the last 4 commits into one. It might be a big commit to review, but could not easily split them while maintaining modularity and making sure each commit passes tests.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -577,23 +792,129 @@ impl NodeInfo {
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let mut builder = TaprootBuilder::new();
```
Would it make sense to have `fn with_capacity(capacity: usize) -> TaprootBuilder` and call it here? I suppose we would need to internally allocate `Vec` with `log2(capacity + 1)` (`size_of::<usize>() * 8 - (capacity + 1).leading_zeros()` if I'm not mistaken)

Kixunil replied:
I don't think we have to optimize for unoptimal trees - they will have bigger problems than a vec reallocating too many times.

sanket1729 replied:
Ah right. capacity is just a good enough guess, does not have to be strict upper bound.

sanket1729 replied:
Done. I don't think there is a +1 here. Used `size_of::<usize>() * 8 - (capacity).leading_zeros()` instead of `capacity + 1`

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -577,23 +792,129 @@ impl NodeInfo {
+    pub leaf: TapLeaf,
     /// The merkle proof (hashing partners) to get this node.
-    merkle_branch: TaprootMerkleBranch,
+    pub merkle_branch: TaprootMerkleBranch,
```
Looks like outside code could break invariants when this is `pub`?

sanket1729 replied:
Indeed, it might break invariants. I remember thinking about this while adding `pub`, but I was mistaken :(

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -603,17 +924,22 @@ impl ScriptLeaf {
+    /// Computes a leaf hash for this [`ScriptLeaf`] if the leaf is known.
     #[inline]
-    pub fn leaf_hash(&self) -> TapLeafHash { TapLeafHash::from_script(&self.script, self.ver) }
+    pub fn leaf_hash(&self) -> Option<TapLeafHash> {
```
If the leaf is not known we already know the hash and should return it?

sanket1729 replied:
I thought it might be confusing because the hidden node might not necessarily be a leaf. It could be a hidden tree.

sanket1729 replied:
I had a note for adding `node_hash()` method that returns the `TapNodeHash`

Kixunil replied:
Ah, that makes sense. Agree with adding `node_hash`.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -712,6 +1046,11 @@ macro_rules! impl_try_from {
+
+impl From<TaprootMerkleBranch> for Vec<sha256::Hash> {
+    fn from(branch: TaprootMerkleBranch) -> Self { branch.0 }
+}
```
FTR, this is still useful even if we drop `into = ...`

Kixunil reviewed (commented):
First pass. I think we should get #1479 in first.

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -521,7 +555,188 @@ impl Default for TaprootBuilder {
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+#[cfg_attr(feature = "serde", serde(into = "NodeInfo"))]
```
It does not. It is still serialized as a new struct. Removing this line results in 
```
thread 'taproot::test::test_merkle_branch_serde' panicked at 'expected Token::Seq { len: Some(2) } but serialized as NewtypeStruct { name: "TaprootMerkleBranch", }
```

Kixunil replied:
Oh, yeah, serializers should treat it as transaparent but they don't have to. Maybe `#[serde(transparent)]` would work? I'm not sure if it'd clash with `try_from` though.

sanket1729 replied:
Yeah, `#[serde(transparent)]` conflicts with `try_from`

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -521,7 +555,188 @@ impl Default for TaprootBuilder {
+
+impl core::hash::Hash for TapTree {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) { self.0.hash(state) }
+}
```
Perhaps, I should add some documentation to this. For taproot trees, because the leaves get sorted by hash while constructing the root, a tree with root A, left child B and right child C is the same as root A with left child B and right child A. 

We had `Eq` reflect this property in `TapTree`, so I thought `Hash` should reflect this property. The `Eq` property is really handy while round-tripping psbts.

Kixunil replied:
If `Eq` returns `true` then comparing hashes *must* also return `true`.

I'm confused by this explanation:

> a tree with root A, left child B and right child C is the same as root A with left child B and right child A.

Maybe you meant to say this?

> a tree with root A, left child B and right child C is the same as root A with left child C and right child B.

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -521,7 +555,188 @@ impl Default for TaprootBuilder {
+        // The builder algorithm invariant guarantees that is_complete builder
+        // have only 1 element in branch and that is not None.
+        // We make sure that we only allow is_complete builders via the from_inner
+        // constructor
```
My bad here

sanket1729 reviewed (commented):
I have a few fixes planned for the PR. Basically, there will be two iterators. 

1) Iterator from `NodeInfo` that returns `ScriptLeaves`.
2) Iterator from `TapTree` that returns a tuple (depth:u8, ver: LeafVersion, script: &Script)

I am also not thrilled with the name `ScriptLeaf` because it also now has hidden nodes, but couldn't think of anything better

Kixunil commented:
> I am also not thrilled with the name `ScriptLeaf` because it also now has hidden nodes

Is there any problem with `Node`? (and call iterator `Nodes`, method `nodes()`...)

sanket1729 commented:
> Is there any problem with Node? (and call iterator Nodes, method nodes()...)

Nodes might represent an internal node too. We are only dealing with tree leaves over here. But they may not be Scripts. So, `ScriptLeaf` is not really appropriate. I wanted to use `TapLeaf`, we are using it for the bare enum without the Merkle path caching. Perhaps, `TapLeafInfo`?

Kixunil commented:
My understanding is that hidden node could potentially hide subtree rather than a single leaf so then "Leaf" depend on whether you define it as "leaf of the whole tree" or "leaf of the known part of the tree". But I admit I don't see a good way to express the latter.

`LeafNode` maybe?

sanket1729 commented:
Sorry for the delay. This is ready for review again.

sanket1729 commented:
If there are no significant changes required to get this one in, we should try to get this in 0.30

apoelstra commented on bitcoin/src/taproot.rs:
```diff
@@ -442,6 +466,37 @@ impl TaprootBuilder {
+        } else {
+            // It is annoying to get rid of this unreachable while maintaining the gurantee that
+            // builder should not change if there is an error
+            unreachable!("Size checked above. Builder guarantees the last element is Some")
```
In 5bf4ca56221dc41488ca8e54aa64351be5ff0471:

This is fine, but FWIW if `self.branch.pop` returns `None`, then it didn't change the builder. So you could just return an error here. (But I don't think you should, since this branch is actually unreachable.)

Also nit: `guarantee` is spelled incorrectly.

sanket1729 replied:
The message was about why it pop cannot return `Some(None)`

apoelstra commented on bitcoin/src/taproot.rs:
```diff
@@ -524,116 +579,158 @@ impl Default for TaprootBuilder {
+
+// TODO: Enforce this in type system in a later PR. Users should not need to unwrap
+// as script leaf when there are no hidden nodes.
+// An idea can be to have the iterator return a tuple (depth, script, version).
```
In 5bf4ca56221dc41488ca8e54aa64351be5ff0471:

I think this TODO is very confusingly worded. I spent some time writing a comment here that you're wrong, but now I understand what you're getting at. I suggest the text
```
// TODO: because a `TapTree` cannot be constructed if it has any hidden nodes, this iterator is
// guaranteed to return script leaves. However, it yields the `LeafNode` type, which may
// represent a hidden node, forcing users to unwrap to get the script. Maybe we should return
// the script (along with depth, version, etc) directly?
```

This is just a nit, but I did spend a bunch of time on it.

apoelstra reviewed (dismissed):
ACK af0b2d04252b71a1b9e5d7c85d6e737e94e51f0e

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -520,26 +573,214 @@ impl Default for TaprootBuilder {
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+            ScriptLeaf::from_leaf_node(self.leaf_iter.next()?)
+        }
```
If you're going to modify this you might as well fix indentation here.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -520,26 +573,214 @@ impl Default for TaprootBuilder {
+
+impl HiddenNodes {
+    /// Converts error into the original incomplete [`NodeInfo`] instance.
+    pub fn into_builder(self) -> NodeInfo {
```
The method is named `into_builder` but returns `NodeInfo`. That doesn't sound right.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -442,6 +464,37 @@ impl TaprootBuilder {
+        } else {
+            // It is annoying to get rid of this unreachable while maintaining the gurantee that
+            // builder should not change if there is an error
+            unreachable!("Size checked above. Builder guarantees the last element is Some")
```
For the future: could we factor out the last element (store it separately), so that the guarantee is represented in type?

sanket1729 replied:
Great idea.

sanket1729 replied:
Indeed. I think this should be possible.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -112,7 +118,12 @@ impl TapNodeHash {
+    }
+
+    /// Assumes the given 32 byte array as hidden [`TapNodeHash`].
+    pub fn assume_hidden(hash: [u8; 32]) -> TapNodeHash {
```
Given a method for constructing from array exists, what is the purpose of this? Just better readability? Perhaps write a note about it in the doc comment?

Kixunil reviewed (commented):
Medium-focus review, I want to re-read this later when I feel better.

sanket1729 commented:
I have updated the description with the changes in this PR. This should also make it easy for writing changelogs.

apoelstra reviewed (approved):
ACK 55f4ea6e3a3d547817e5c799e47e442aad77cbe0

tcharding reviewed (dismissed):
ACK ae2d154ded2bfbe388e7ded53c2b0e9e384f5af9

My ack is given without very deep knowledge in this area, perhaps don't count it towards the two required. I had a bunch of nits around docs but since we are so close to release I think this is ok like this and I'll come back and do a PR against all the docs module wide.

tcharding commented:
@Kixunil this one needs your attention please bro.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -987,6 +1004,46 @@ impl LeafNode {
+    /// The script.
+    script: &'leaf Script,
+    /// The merkle proof (hashing partners) to get this node.
+    merkle_branch: TaprootMerkleBranch,
```
Shouldn't this be borrowed as well?

sanket1729 replied:
Indeed, I don't know what I was thinking here.

Kixunil reviewed (approved):
ACK ae2d154ded2bfbe388e7ded53c2b0e9e384f5af9

The suggested improvements can go to separate PR.

apoelstra commented:
Unfortunately this needs a rebase. The merge commit does not compile with 1.41.

Kixunil commented:
Since the reviews are going to be dismissed anyway @sanket1729 please consider addressing my comments in this PR

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -524,116 +583,158 @@ impl Default for TaprootBuilder {
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+#[cfg_attr(feature = "serde", serde(into = "NodeInfo"))]
+#[cfg_attr(feature = "serde", serde(try_from = "NodeInfo"))]
+pub struct TapTree(pub(crate) NodeInfo);
```
Good catch.

sanket1729 commented on bitcoin/src/psbt/serialize.rs:
```diff
@@ -385,23 +386,21 @@ impl Deserialize for (Vec<TapLeafHash>, KeySource) {
-            }
-        // This should be unreachable as we Taptree is already finalized
-            _ => unreachable!(),
+        let mut buf = Vec::new();
```
Done, implemented using exact size iter

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -520,6 +520,131 @@ impl Default for TaprootBuilder {
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct TapTree(pub(crate) TaprootBuilder);
```
Good catch. Fixed in the third commit that actually fixes this invariant. The first commit was intended to be move-only. This is removed when we change to `TapTree(NodeInfo)`

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -357,6 +377,14 @@ impl TaprootBuilder {
     pub fn new() -> Self { TaprootBuilder { branch: vec![] } }
 
+    /// Creates a new instance of [`TaprootBuilder`] with a capacity hint for `size` elements.
+    /// The size here should be maximum depth of the tree.
```
So, the idea here is that first line should be brief description and details should follow in the next paragh.

Kixunil replied:
Yes, rustdoc then formats them appropriately.

sanket1729 commented:
@Kixunil addressed all the comments. This is ready for review. 

@apoelstra Great that we have the test scripts to test these kinds of silent merge failures in these kinds of long-range PRs.

tcharding reviewed (approved):
ACK b6214651cb7579a5773f3da062b620b92d8f74e6

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -700,48 +814,184 @@ impl NodeInfo {
+        S: serde::Serializer,
+    {
+        use serde::ser::SerializeSeq;
+        let mut seq = serializer.serialize_seq(Some(self.leaves.len()))?;
```
I think this needs to be multiplied by 2 since there are two calls of `serialize_element`.

sanket1729 replied:
good catch

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -700,48 +814,184 @@ impl NodeInfo {
+                A: serde::de::SeqAccess<'de>,
+            {
+                let size = seq.size_hint()
+                    .map(|x| core::mem::size_of::<usize>()*8 - x.leading_zeros() as usize)
```
`x` needs to be divided by 2.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -724,25 +724,43 @@ impl TryFrom<NodeInfo> for TapTree {
-impl<'tree> ExactSizeIterator for TapTreeIter<'tree> {
+impl<'tree> ExactSizeIterator for LeafNodes<'tree> {
     #[inline]
     fn len(&self) -> usize { self.leaf_iter.len() }
```
Oh, just realized this is wrong because it has to forward `size_hint` correctly. This method also doesn't need to be implemented.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -724,25 +724,43 @@ impl TryFrom<NodeInfo> for TapTree {
+    type Item = &'a LeafNode;
 
     #[inline]
     fn next(&mut self) -> Option<Self::Item> { self.leaf_iter.next() }
```
Missing `size_hint`.

Kixunil commented on bitcoin/src/taproot.rs:
```diff
@@ -724,25 +724,43 @@ impl TryFrom<NodeInfo> for TapTree {
 }
 
-impl<'tree> ExactSizeIterator for TapTreeIter<'tree> {
+impl<'tree> ExactSizeIterator for LeafNodes<'tree> {
```
Other traits for consideration here and above: `FusedIterator`, `DoubleEndedIterator`

Kixunil reviewed (commented):
Found some bugs unfortunately.

sanket1729 commented on bitcoin/src/taproot.rs:
```diff
@@ -263,7 +278,14 @@ impl TaprootSpendInfo {
+            } else {
+                // As of now, we do not store any information about hidden nodes in
+                // TaprootSpendInfo.
+                continue;
```
This is not a bug. `TaprootSpendInfo` is a cached data structure for storing information about spending leaves. As of now, we don't cache anything for hidden leaves.

tcharding replied:
Perhaps this is cleaner:
```rust
    pub fn from_node_info<C: secp256k1::Verification>(
        secp: &Secp256k1<C>,
        internal_key: UntweakedPublicKey,
        node: NodeInfo,
    ) -> TaprootSpendInfo {
        // Create as if it is a key spend path with the given merkle root
        let root_hash = Some(node.hash);
        let mut info = TaprootSpendInfo::new_key_spend(secp, internal_key, root_hash);

        for leaves in node.leaves {
            match leaves.leaf {
                TapLeaf::Hidden(_) => {
                    // We don't store any information about hidden nodes in TaprootSpendInfo.
                }
                TapLeaf::Script(script, ver) => {
                    let key = (script, ver);
                    let value = leaves.merkle_branch;
                    match info.script_map.get_mut(&key) {
                        None => {
                            let mut set = BTreeSet::new();
                            set.insert(value);
                            info.script_map.insert(key, set);
                        },
                        Some(set) => {
                            set.insert(value);
                        }
                    }
                },
            }
        }
        info
    }
```

sanket1729 commented:
> Found some bugs unfortunately.

@Kixunil, good eye and great work on catching those bugs on `*2` and `/2`. Surprisingly easy to miss and overlook. Force pushed fixes

tcharding commented:
Ouch, that was a hard review to read @Kixunil after my ack. You are a good bloke to have on the team.

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -521,26 +576,238 @@ impl Default for TaprootBuilder {
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for IncompleteBuilder {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use self::IncompleteBuilder::*;
```
OT: These `use self::<ErrorType>::*` statements are sprinkled around the codebase because I was coding sometime with only half my brain connected, the `self::` is totally useless.

tcharding commented on bitcoin/src/psbt/serialize.rs:
```diff
@@ -384,23 +387,22 @@ impl Deserialize for (Vec<TapLeafHash>, KeySource) {
-            }
-        // This should be unreachable as we Taptree is already finalized
-            _ => unreachable!(),
+        let capacity = self.script_leaves().map(|l|
+                l.merkle_branch().len()* 32 + // merkle branch
+                l.script().len() + VarInt(l.script().len() as u64).len() + // script version
+                1 // leaf version
+            ).sum::<usize>();
```
Perhaps this is mildly cleaner:
```rust
        let capacity = self.script_leaves().map(|l| {
            l.merkle_branch().len() * 32 // merkle branch
                + l.script().len() + VarInt(l.script().len() as u64).len() // script version
                + 1 // leaf version
        }).sum::<usize>();
```

tcharding commented on bitcoin/src/psbt/serialize.rs:
```diff
@@ -384,23 +387,22 @@ impl Deserialize for (Vec<TapLeafHash>, KeySource) {
+            //
+            // TaprootMerkleBranch can only have len atmost 128(TAPROOT_CONTROL_MAX_NODE_COUNT).
+            // safe to cast from usize to u8
+            buf.push(leaf_info.merkle_branch().len() as u8);
```
Why the `* 32` when we push just the length here?

sanket1729 replied:
another good catch. I really need to think before I am typing :(

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -443,6 +471,33 @@ impl TaprootBuilder {
+    /// Converts the builder into a [`NodeInfo`] if the builder is a full tree with possibly
+    /// hidden nodes
+    ///
+    /// # Errors:
```
Its not important right now but just FYI; we don't need the `:` here.

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -443,6 +471,33 @@ impl TaprootBuilder {
+    ///
+    /// # Errors:
+    ///
+    /// IncompleteBuilder::NotFinalized if the builder is not finalized. The builder
```
```suggestion
    /// [`IncompleteBuilder::NotFinalized`] if the builder is not finalized. The builder
```

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -521,26 +576,238 @@ impl Default for TaprootBuilder {
+
+    /// Constructs [`TapTree`] from a [`TaprootBuilder`] if it is complete binary tree.
+    ///
+    /// # Returns
+    /// A [`TapTree`] iff the `builder` is complete, otherwise return [`IncompleteBuilder`]
```
nit: Missing newline after `# Returns`.

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -521,26 +576,238 @@ impl Default for TaprootBuilder {
+
+    /// Constructs [`TapTree`] from a [`NodeInfo`] if it is complete binary tree.
+    ///
+    /// # Returns
```
nit: And here.

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -521,26 +576,238 @@ impl Default for TaprootBuilder {
+            ScriptLeaf::from_leaf_node(self.leaf_iter.next()?)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) { self.leaf_iter.size_hint() }
```
Any reason `next` get inlined but not `size_hint`?

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -567,69 +836,256 @@ impl NodeInfo {
         })
     }
+
+    /// Iterator over all leaves(including hidden leaves) in the tree.
```
```suggestion
    /// Creates an iterator over all leaves (including hidden leaves) in the tree.
```

tcharding commented on bitcoin/src/taproot.rs:
```diff
@@ -567,69 +836,256 @@ impl NodeInfo {
     }
 
-    /// Computes a leaf hash for this [`ScriptLeaf`].
+    /// Computes a leaf hash for this [`ScriptLeaf`] if the leaf is known.
+    /// See [`LeafNode::node_hash`] for computing the [`TapNodeHash`] which returns
+    /// the hidden node hash if the node is hidden.
+    /// This [`TapLeafHash`] is useful while signing taproot script spends.
```
Perhaps

```suggestion
    /// Computes a leaf hash for this [`ScriptLeaf`] if the leaf is known.
    ///
    /// This [`TapLeafHash`] is useful while signing taproot script spends.
    ///
    /// See [`LeafNode::node_hash`] for computing the [`TapNodeHash`] which returns the hidden node
    /// hash if the node is hidden.
```

tcharding reviewed (approved):
I tried to put a bit more review effort in, many suggestions are only mildly better, feel free to ignore.

And just to be clear, I'm not a domain expert on taproot, take my review as code/style only please.

sanket1729 commented:
@tcharding, thanks for the detailed review. Force pushed addressing your comments.

tcharding reviewed (approved):
ACK 74022baa44d7efbe3e0ffc58126dc7dfc9b42344

tcharding commented:
Review needed on this one please crew, this is the last big PR to go in before we can consider release (other than formatting changes and the hashes/hex debate). Thanks.

apoelstra commented:
Sorry @tcharding I am ready to ACK this, but I spent all day yesterday trying to review it using my new nix setup, and running into unimplemented stuff that I had to fix (with pretty slow iteration times, and a *long* time spent trawling through cargo source code). You can see my efforts [on the crate2nix repo PR list](https://github.com/kolloch/crate2nix/pulls).

apoelstra reviewed (approved):
ACK 74022baa44d7efbe3e0ffc58126dc7dfc9b42344
